src/python/pants/backend/python/mixed_interpreter_constraints/py_constraints_test.py

```diff
@@ -97,8 +100,8 @@ def test_constraints_summary(rule_runner: RuleRunner) -> None:
         Target,Constraints,Transitive Constraints,# Dependencies,# Dependees\r
         app,CPython==3.7.*,"CPython==2.7.*,==3.7.*,>=3.6 OR CPython==3.7.*,>=3.5,>=3.6",3,0\r
         lib1,CPython==2.7.* OR CPython>=3.5,CPython==2.7.* OR CPython>=3.5,0,1\r
-        lib2,CPython>=3.6,CPython>=3.6,2,0\r
-        lib2/a.py,CPython>=3.6,CPython>=3.6,2,3\r
-        lib2/b.py,CPython>=3.6,CPython>=3.6,2,3\r
+        lib2,CPython>=3.6,CPython>=3.6,0,0\r
+        lib2/a.py,CPython>=3.6,CPython>=3.6,0,2\r
+        lib2/b.py,CPython>=3.6,CPython>=3.6,0,2\r
```

Comment:
This was tough to figure out. `lib2/a.py` and `lib2/b.py` no longer depend on each other because dependency inference is now activated and dependencies aren't added on sibling files.

`lib2/` is weirder. It should be claiming 2 dependencies: on `lib2/a.py` and `lib2/b.py`. But it doesn't because our transitive targets code expands the "root" so that it becomes `lib2/a.py`, rather than `lib2`:

https://github.com/pantsbuild/pants/blob/e02b59fdbb6b21fb203459467e780b21c3cb80a9/src/python/pants/engine/internals/graph.py#L276-L280

I added print statements to `py_constraints.py` to first print the address of `all_python_targets`, which includes generators and generated:

```python
[Address(app), Address(lib1), Address(lib2), Address(lib2/a.py), Address(lib2/b.py)]
```

Good so far. But then I printed the `(transitive_targets.roots, transitive_targets.dependencies)` per tgt:

```
app [Address(lib2/a.py), Address(lib2/b.py), Address(lib1)]
lib1 []
lib2/a.py []
lib2/a.py []
lib2/b.py []
```

This has been the behavior for a long time apparently. https://github.com/pantsbuild/pants/pull/10511. It's fixed by always using `UnexpandedTargets` for `roots_as_targets` in the above transitive targets rule. That feels right to change that way, but I'm having a hard time understanding this all.

Reply:
> This has been the behavior for a long time apparently. #10511. It's fixed by always using `UnexpandedTargets` for `roots_as_targets` in the above transitive targets rule. That feels right to change that way, but I'm having a hard time understanding this all.

Yea, I think that it makes sense in general for `TransitiveTargets` to give you back your roots verbatim, and to in general not expand targets as it iterates.

Reply:
Cool, will try to fix in a followup change..going to keep this PR using the original semantics.